File: demos/common/src/ie_config_helper.cpp
// Copyright (C) 2020 Intel Corporation
// SPDX-License-Identifier: Apache-2.0
//

#include "samples/ie_config_helper.hpp"
#include "samples/args_helper.hpp"

std::string formatDeviceString(const std::string& deviceString) {
    std::string formattedString = deviceString;
    int changeCase = 1;

    for (size_t i = 0; i < deviceString.size(); ++i) {
        // For the special case of MYRIAD device names in "MYRIAD.0.0-ma0000" format
        std::string prefix = deviceString.substr(i, 2);
        if (prefix == "ma") {
            changeCase = 0;
        } else if (prefix == "MA" || prefix == "Ma" || prefix == "mA") {
            changeCase = -1;
        } else if (changeCase != 1 && deviceString[i] == ',') {
            changeCase = 1;
        }

        if (changeCase == 1) {
            formattedString[i] = std::toupper(deviceString[i]);
        } else if (changeCase == -1) {
            formattedString[i] = std::tolower(deviceString[i]);
        }
    }

    return formattedString;
}

std::map<std::string, std::string> createConfig(const std::string& deviceString,
                                                const std::string& nstreamsString,
                                                int nthreads,
                                                bool minLatency) {
    std::map<std::string, std::string> config;

    std::set<std::string> devices;
    for (const std::string& device : parseDevices(deviceString)) {
        devices.insert(device);
    }
    std::map<std::string, unsigned> deviceNstreams = parseValuePerDevice(devices, nstreamsString);
    
    for (auto& device : devices) {
        if (device == "CPU") {  // CPU supports a few special performance-oriented keys
            if (minLatency) {
                config.insert({ CONFIG_KEY(CPU_THROUGHPUT_STREAMS), "1" });
                continue;
            }
            
            // limit threading for CPU portion of inference
            if (nthreads != 0) {
                config.insert({ CONFIG_KEY(CPU_THREADS_NUM), std::to_string(nthreads) });
            }

            if (deviceString.find("MULTI") != std::string::npos && devices.find("GPU") != devices.end()) {
                config.insert({ CONFIG_KEY(CPU_BIND_THREAD), CONFIG_VALUE(NO) });
            } else {
                // pin threads for CPU portion of inference
                config.insert({ CONFIG_KEY(CPU_BIND_THREAD), CONFIG_VALUE(YES) });
            }

            // for CPU execution, more throughput-oriented execution via streams
            config.insert({ CONFIG_KEY(CPU_THROUGHPUT_STREAMS), (deviceNstreams.count(device) > 0
                                                                 ? std::to_string(deviceNstreams.at(device))
                                                                 : CONFIG_VALUE(CPU_THROUGHPUT_AUTO)) });
        } else if (device == "GPU") {
            if (minLatency) {
                config.insert({ CONFIG_KEY(GPU_THROUGHPUT_STREAMS), "1" });
                continue;
            }

            config.insert({ CONFIG_KEY(GPU_THROUGHPUT_STREAMS), (deviceNstreams.count(device) > 0
                                                                 ? std::to_string(deviceNstreams.at(device))
                                                                 : CONFIG_VALUE(GPU_THROUGHPUT_AUTO)) });

            if (deviceString.find("MULTI") != std::string::npos && devices.find("CPU") != devices.end()) {
                // multi-device execution with the CPU + GPU performs best with GPU throttling hint,
                // which releases another CPU thread (that is otherwise used by the GPU driver for active polling)
                config.insert({ CLDNN_CONFIG_KEY(PLUGIN_THROTTLE), "1" });
            }
        } else if (device.find("MYRIAD") == 0) {
            if (minLatency) {
                config.insert({ InferenceEngine::MYRIAD_THROUGHPUT_STREAMS, "1" });
                continue;
            }

            if (deviceNstreams.count(device) > 0) {
                config.insert({ InferenceEngine::MYRIAD_THROUGHPUT_STREAMS,
                                std::to_string(deviceNstreams.at(device)) });
            }
        }
    }

    return config;
}

std::map<std::string, std::string> createDefaultConfig(std::string deviceString) {
    std::map<std::string, std::string> config;

    if (deviceString.find("MYRIAD") != std::string::npos) {
        config.insert({ InferenceEngine::MYRIAD_THROUGHPUT_STREAMS, "1" });
    }
    
    return config;
}

File: demos/python_demos/common/ie_config_helper.py
"""
 Copyright (C) 2020 Intel Corporation

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
"""

def format_device_string(device_string):
    formatted_string = ''
    change_case = 1

    for i in range(len(device_string)):
        # For the special case of MYRIAD device names in "MYRIAD.0.0-ma0000" format
        prefix = device_string[i:i+2]
        if prefix == 'ma':
            change_case = 0
        elif prefix == 'MA' or prefix == 'Ma' or prefix == 'mA':
            change_case = -1
        elif change_case != 1 and device_string[i] == ',':
            change_case = 1

        if change_case == 1:
            formatted_string += device_string[i].upper()
        elif change_case == -1:
            formatted_string += device_string[i].lower()
        elif change_case == 0:
            formatted_string += device_string[i]

    return formatted_string


def create_config(device_string, nstreams_string, nthreads, min_latency=False):
    config = {}

    is_multi = (device_string.find('MULTI') == 0)
    devices = device_string.split(':')[-1].split(',')

    devices_nstreams = {}
    if nstreams_string:
        devices_nstreams = {device: nstreams_string for device in devices if device in device_string} \
                                    if nstreams_string.isdigit() \
                                    else dict([device.split(':') for device in nstreams_string.split(',')])

    for device in devices:
        if device == 'CPU':
            if min_latency:
                config['CPU_THROUGHPUT_STREAMS'] = '1'
                continue

            if nthreads:
                config['CPU_THREADS_NUM'] = str(nthreads)
            
            config['CPU_BIND_THREAD'] = 'NO' if is_multi and 'GPU' in devices \
                                             else 'YES'

            if 'CPU' in devices_nstreams:
                config['CPU_THROUGHPUT_STREAMS'] = devices_nstreams['CPU'] if int(devices_nstreams['CPU']) > 0 \
                                                                           else 'CPU_THROUGHPUT_AUTO'
        elif device == 'GPU':
            if min_latency:
                config['GPU_THROUGHPUT_STREAMS'] = '1'
                continue

            if 'GPU' in devices_nstreams:
                config['GPU_THROUGHPUT_STREAMS'] = devices_nstreams['GPU'] if int(devices_nstreams['GPU']) > 0 \
                                                                           else 'GPU_THROUGHPUT_AUTO'
            
            if is_multi and 'CPU' in devices:
                config['PLUGIN_THROTTLE'] = '1'
        elif 'MYRIAD' in device:
            if min_latency:
                config['MYRIAD_THROUGHPUT_STREAMS'] = '1'
                continue

            if device in devices_nstreams:
                config['MYRIAD_THROUGHPUT_STREAMS'] = devices_nstreams[device]

    return config


def create_default_config(device_string):
    config = {}

    if 'MYRIAD' in device_string:
        config['MYRIAD_THROUGHPUT_STREAMS'] = '1'

    return config
